### PR TITLE
improve draining

### PR DESCRIPTION
**Pull request by xh3b4sd**
## Checklist

- [x] Update changelog in CHANGELOG.md.

## Context

I found an issue during our Team Alerts Session where the latest version of `aws-operator` has issues cleaning up after draining. I made three changes in here which are somehow interwoven. Still need to test this. 

```
{
    "caller": "github.com/giantswarm/operatorkit/controller/controller.go:511",
    "controller": "aws-operator-drainer-controller",
    "event": "update",
    "level": "error",
    "loop": "31",
    "message": "failed processing event",
    "object": "/apis/infrastructure.giantswarm.io/v1alpha2/namespaces/default/awsmachinedeployments/p5efj",
    "stack" {
        "annotation": "ValidationError: No active Lifecycle Action found with instance ID i-08406e13ee788fc10\n\tstatus code: 400, request id: b015a014-87bd-4496-80ad-4b57af675c25",
        "kind": "unknown",
        "stack" [
            {
                "file": "/root/project/service/controller/resource/drainerfinalizer/resource.go",
                "line": 90
            },
            {
                "file": "/root/project/service/controller/resource/drainerfinalizer/resource.go",
                "line": 198
            },
            {
                "file": "/root/project/service/controller/resource/drainerfinalizer/create.go",
                "line": 12
            },
            {
                "file": "/go/pkg/mod/github.com/giantswarm/operatorkit@v0.2.0/resource/wrapper/retryresource/basic_resource.go",
                "line": 52
            },
            {
                "file": "/go/pkg/mod/github.com/giantswarm/backoff@v0.2.0/retry.go",
                "line": 23
            },
            {
                "file": "/go/pkg/mod/github.com/giantswarm/operatorkit@v0.2.0/resource/wrapper/retryresource/basic_resource.go",
                "line": 64
            },
            {
                "file": "/go/pkg/mod/github.com/giantswarm/operatorkit@v0.2.0/resource/wrapper/metricsresource/basic_resource.go",
                "line": 43
            },
            {
                "file": "/go/pkg/mod/github.com/giantswarm/operatorkit@v0.2.0/controller/controller.go",
                "line": 617
            }
        ]
    },
    "time": "07:41:32",
    "version": "206959606"
}
```

### CODE REVIEW

File: service/controller/control_plane_drainer_resource_set.go [hunk 1]
```diff
@@ -177,7 +177,8 @@ func newControlPlaneDrainerResourceSet(config controlPlaneDrainerResourceSetConf
 
 func controlPlaneDrainerLabelMapFunc(cr metav1.Object) map[string]string {
 	return map[string]string{
-		label.Cluster:      key.ClusterID(cr),
-		label.ControlPlane: key.ControlPlaneID(cr),
+		label.Cluster:         key.ClusterID(cr),
+		label.ControlPlane:    key.ControlPlaneID(cr),
+		label.OperatorVersion: key.OperatorVersion(cr),
```
Comment (xh3b4sd) [inline service/controller/control_plane_drainer_resource_set.go:182]: Also preparing the Control Plane controller in advance.

File: service/controller/machine_deployment_drainer_resource_set.go [hunk 2]
```diff
@@ -179,5 +179,6 @@ func machineDeploymentDrainerLabelMapFunc(cr metav1.Object) map[string]string {
 	return map[string]string{
 		label.Cluster:           key.ClusterID(cr),
 		label.MachineDeployment: key.MachineDeploymentID(cr),
+		label.OperatorVersion:   key.OperatorVersion(cr),
 	}
```
Comment (xh3b4sd) [inline service/controller/machine_deployment_drainer_resource_set.go:183]: Adding the operator version label here so that only operators actually creating the `DrainerConfig` CRs drain and cleanup their own CRs. The map is configured to be used for both `drainerinitializer` and `drainerfinalizer` handlers.

File: service/controller/resource/drainerfinalizer/error.go [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+	"regexp"
 
 	"github.com/giantswarm/microerror"
 )
 
+var (
+	// noActiveLifeCycleActionRegExp is a fuzzy regular expression to match
+	// Autoscaling errors which we have to string match due to the lack of proper
+	// error types in the AWS SDK.
+	noActiveLifeCycleActionRegExp = regexp.MustCompile(`(?m)[nN][oO].*[lL][iI][fF][eE].*[cC][yY][cC][lL][eE].*[fF][oO][uU][nN][dD]`)
```
Comment (xh3b4sd) [inline service/controller/resource/drainerfinalizer/error.go:13]: This is the dumbest idea I could have come up with. Ideas welcome. It is just that the lower/upper case changed here again and we need some solution to fuzzy match that. Not a regexp hero here.

File: service/controller/resource/drainerfinalizer/resource.go [hunk 4]
```diff
@@ -204,6 +211,15 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
...
 		}
 
+		// Timeout drainer configs should simply be deleted without draining since
+		// the time window for draining did expire at this point.
+		for _, dc := range timeoutDrainerConfigs {
+			err = r.deleteDrainerConfig(ctx, dc)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
```
Comment (xh3b4sd) [inline service/controller/resource/drainerfinalizer/resource.go:221]: I thought this makes sense but would like to hear about more opinions. Now due to the error matcher issues the `DrainerConfig` CRs are not deleted even though they timed out and the instances are gone. Not sure if this may have unforeseen side effects.

File: service/controller/resource/drainerfinalizer/error.go (hunk 3 above)
Comment (kopiczko) [reply service/controller/resource/drainerfinalizer/error.go:13]: You could add `i` flag instead. `(?m)` -> `(?im)`.

https://stackoverflow.com/questions/15326421/how-do-i-do-a-case-insensitive-regular-expression-in-go

Comment (kopiczko) [reply service/controller/resource/drainerfinalizer/error.go:13]: I'm also not sure why you need `m` flag.

File: service/controller/machine_deployment_drainer_resource_set.go (hunk 2 above)
Comment (tuommaki) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: Why would different operators reconcile objects for same cluster? When there's cluster and machine deployment IDs, I'd expect that this already pins it down to single instance? What if cluster gets upgraded and for unknown reason there was `DrainerConfig` for a given node pool - wouldn't the CR get then orphaned?

File: service/controller/resource/drainerfinalizer/resource.go (hunk 4 above)
Comment (tuommaki) [reply service/controller/resource/drainerfinalizer/resource.go:221]: In Azure we actually restart draining when it times out. So far it has worked for us quite well and currently our only problems are with occasional stuck `coredns` and `nginx-ingress-controller` PODs.

Comment (xh3b4sd) [reply service/controller/resource/drainerfinalizer/resource.go:221]: In the case here we would start over forever which renders the timeout quite useless. That is why I thought we should draw a line an some point. Do you think this here is wrong or not so good?

File: service/controller/machine_deployment_drainer_resource_set.go (hunk 2 above)
Comment (xh3b4sd) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: We just create and list `DrainerConfig` CRs. So `8.2.0` would do the finalizing that `8.3.0` would do as well. Now with the additional operator version label the operators only consider their own managed CRs.

Comment (xh3b4sd) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: Upgrade and scaling should not happen together anyway. When upgrade is started the new operator version is reconciling everything of the cluster. Combined with the deletion on timeout this should be fine.

Comment (tuommaki) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: > Upgrade and scaling should not happen together anyway.

Is `cluster-autoscaler` actively disabled during upgrade?

File: service/controller/resource/drainerfinalizer/resource.go (hunk 4 above)
Comment (tuommaki) [reply service/controller/resource/drainerfinalizer/resource.go:221]: > Do you think this here is wrong or not so good?

I think that in general this is the right approach. For different circumstances and different customer requirements different approaches are better, but when you can behave accordingly and draw a strict line to timeout, I think that is good.

TL;DR; When no customer is screaming, I like this approach. :+1:

File: service/controller/machine_deployment_drainer_resource_set.go (hunk 2 above)
Comment (xh3b4sd) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: Not that I am aware of. The responsibility of that app just got recently into Firecracker and we did not have yet the chance to get around to it. Do you think this is necessary? How to do it?

Comment (tuommaki) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: > Do you think this is necessary?

From abandoned CR point of you yes, but from real life usage no.

> How to do it?

No clue. Depends probably from many factors, but in general scaling the deployment to `0 replicas` would do it. Problem is though that then there's tight coupling between an app and provider operator.

Might be useful to check how others deal with it (if they even do). There could be even some labeling trick or something to signal CA that it should leave a given ASG alone until the label is removed.

Comment (kopiczko) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: So it was a bug? Two operators would reconcile the same resource?

Comment (xh3b4sd) [reply service/controller/machine_deployment_drainer_resource_set.go:183]: I think so.

File: service/controller/resource/drainerfinalizer/error.go (hunk 3 above)
Comment (xh3b4sd) [reply service/controller/resource/drainerfinalizer/error.go]: Nice one. That works.

**Comment by xh3b4sd** [comment]
I think this makes sense. We do some more testing together with Spot Instances and HA Masters changes. Going with that and hopefully release tomorrow.
